src/settings.rs

```diff
@@ -498,6 +502,193 @@ pub enum InstallSettingsError {
     InitNotSupported,
+    #[error(transparent)]
+    UrlOrPath(#[from] UrlOrPathError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum UrlOrPathError {
+    #[error("Error parsing URL `{0}`")]
+    Url(String, #[source] url::ParseError),
+    #[error("The specified path `{0}` does not exist")]
+    PathDoesNotExist(PathBuf),
+    #[error("Error fetching URL `{0}`")]
+    Reqwest(Url, #[source] reqwest::Error),
+    #[error("I/O error when accessing `{0}`")]
+    Io(PathBuf, #[source] std::io::Error),
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize, Clone)]
+pub enum UrlOrPath {
+    Url(Url),
+    Path(PathBuf),
+}
+
+impl Display for UrlOrPath {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            UrlOrPath::Url(url) => f.write_fmt(format_args!("{url}")),
+            UrlOrPath::Path(path) => f.write_fmt(format_args!("{}", path.display())),
+        }
+    }
+}
+
+impl FromStr for UrlOrPath {
+    type Err = UrlOrPathError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match Url::parse(s) {
+            Ok(url) => Ok(UrlOrPath::Url(url)),
+            Err(url::ParseError::RelativeUrlWithoutBase) => {
+                // This is most likely a relative path (`./boop` or `boop`)
+                // or an absolute path (`/boop`)
+                //
+                // So we'll see if such a path exists, and if so, use it
+                let path = PathBuf::from(s);
+                if path.exists() {
+                    Ok(UrlOrPath::Path(path))
+                } else {
+                    Err(UrlOrPathError::PathDoesNotExist(path))
+                }
+            },
+            Err(e) => Err(UrlOrPathError::Url(s.to_string(), e)),
+        }
+    }
+}
+
+#[cfg(feature = "cli")]
+impl clap::builder::TypedValueParser for UrlOrPath {
+    type Value = UrlOrPath;
+
+    fn parse_ref(
+        &self,
+        cmd: &clap::Command,
+        _arg: Option<&clap::Arg>,
+        value: &std::ffi::OsStr,
+    ) -> Result<Self::Value, clap::Error> {
+        let value_str = value.to_str().ok_or_else(|| {
+            let mut err = clap::Error::new(clap::error::ErrorKind::InvalidValue);
+            err.insert(
+                ContextKind::InvalidValue,
+                ContextValue::String(format!("`{value:?}` not a UTF-8 string")),
+            );
+            err
+        })?;
+        match UrlOrPath::from_str(value_str) {
+            Ok(v) => Ok(v),
+            Err(from_str_error) => {
+                let mut err = clap::Error::new(clap::error::ErrorKind::InvalidValue).with_cmd(cmd);
+                err.insert(
+                    clap::error::ContextKind::Custom,
+                    clap::error::ContextValue::String(from_str_error.to_string()),
+                );
+                Err(err)
+            },
+        }
+    }
+}
+
+#[test]
+fn url_or_path_parses() -> Result<(), Box<dyn std::error::Error>> {
```

Comment:
Is there a reason you didn't put this inside a `mod tests`?

Reply:
IMHO, the benefit is that the tests are easier to discover, rather than having to either grep for `#[test]` or happen upon the test sandwiched between other code. But it's not a big deal, just wondering if there was a reason :)

Reply:
It's competing priorities between "Keep code that tests code near the code" and "Keep test code together" 😄